File: app/src/main/java/com/example/mlallemant/mentalbattle/UI/Menu/Fragment/SessionLauncherFragment.java
package com.example.mlallemant.mentalbattle.UI.Menu.Fragment;

import android.app.Fragment;
import android.app.FragmentManager;
import android.app.FragmentTransaction;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.example.mlallemant.mentalbattle.R;
import com.example.mlallemant.mentalbattle.UI.Menu.MenuActivity;
import com.example.mlallemant.mentalbattle.UI.Session.SessionActivity;
import com.example.mlallemant.mentalbattle.Utils.DatabaseManager;
import com.example.mlallemant.mentalbattle.Utils.Player;
import com.example.mlallemant.mentalbattle.Utils.Session;
import com.example.mlallemant.mentalbattle.Utils.Utils;

import java.util.ArrayList;
import java.util.List;

import static com.facebook.FacebookSdk.getApplicationContext;

/**
 * Created by m.lallemant on 15/11/2017.
 */

public class SessionLauncherFragment extends Fragment {

    private final static String TAG = "SessionLauncherFragment";

    //UI
    private ImageView iv_back;
    private TextView tv_name;
    private ImageView iv_list_view;
    private TextView tv_nb_players;
    private TextView tv_nb_players_title;
    private Button btn_launch_ready;
    private ListView lv_players;

    //Utils
    private Player currentPlayer;
    private DatabaseManager db;
    private boolean isCreator = false;
    private String name;
    private String password;
    private Session session;
    private int nbPlayer = 0;
    private boolean isReady = false;
    private ArrayList<Player> playerModel;
    private PlayersAdapter adapter;
    private boolean isEverybodyReady = false;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.menu_session_fragment, container, false);

        MenuActivity menuActivity = (MenuActivity) getActivity();

        currentPlayer = menuActivity.getCurrentPlayer();

        isCreator = getArguments().getBoolean("creator");
        name = getArguments().getString("name");
        password = getArguments().getString("password");
        session = new Session(name, password, Utils.SESSION_STATE_WAITING);

        db = DatabaseManager.getInstance();
        db.initListenerCurrentSession(session);

        initUI(v);
        initListener();

        menuActivity.setOnBackPressedListener(new MenuActivity.OnBackPressedListener() {
            @Override
            public void doBack() {
                if (db != null) {
                    quitSession();
                }
            }
        });

        return v;
    }


    private void initUI(View v) {
        iv_back = (ImageView) v.findViewById(R.id.iv_session_back);
        tv_name = (TextView) v.findViewById(R.id.tv_session_name);
        iv_list_view = (ImageView) v.findViewById(R.id.iv_session_list_view);
        tv_nb_players = (TextView) v.findViewById(R.id.tv_session_nb_player);
        tv_nb_players_title = (TextView) v.findViewById(R.id.tv_session_nb_player_title);
        btn_launch_ready = (Button) v.findViewById(R.id.btn_session_launch_ready);
        lv_players = (ListView) v.findViewById(R.id.lv_session_players);

        lv_players.setVisibility(View.VISIBLE);
        iv_list_view.setVisibility(View.INVISIBLE);

        playerModel = new ArrayList<>();
        adapter = new PlayersAdapter(playerModel, session,isCreator, getActivity());
        lv_players.setAdapter(adapter);
        adapter.clear();

        if(isCreator) {
            btn_launch_ready.setText("LAUNCH");
        } else {
            btn_launch_ready.setText("READY ?");
        }

        nbPlayer++;
        tv_nb_players.setText(""+nbPlayer);
    }


    private void initListener(){
        iv_back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                db.removeListenerCurrentSession(session);
                db.insertPlayerInLobby(currentPlayer);
                if (isCreator) {
                    db.deleteSession(session);
                    returnSelectorFragment();
                } else {
                    quitSession();
                    db.insertPlayerInLobby(currentPlayer);
                }

            }
        });

        btn_launch_ready.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (isCreator) {
                    if (isEverybodyReady) {
                        db.updateStateSession(session, Utils.SESSION_STATE_LAUNCH_PARTY);
                    }else {
                        makeToast("All the players are not ready");
                    }
                } else {
                    if (!isReady) {
                        isReady = true;
                        btn_launch_ready.setText("READY !");
                        btn_launch_ready.setBackground(ContextCompat.getDrawable(getActivity(), R.drawable.shape_green));
                        db.updatePlayerReady(session, currentPlayer, Utils.SESSION_RDY_YES);
                    } else {
                        isReady = false;
                        btn_launch_ready.setText("READY ?");
                        btn_launch_ready.setBackground(ContextCompat.getDrawable(getActivity(), R.drawable.shape_orange));
                        db.updatePlayerReady(session, currentPlayer, Utils.SESSION_RDY_NO);
                    }
                }
            }
        });


        db.setOnSessionUpdateListener(new DatabaseManager.OnSessionUpdateListener() {
            @Override
            public void updateSessionUI(Session session_) {

                if (session_ != null) {
                    session = session_;

                    List<Player> players = session_.getPlayerList();
                    adapter.clear();
                    adapter.addAll(players);
                    tv_nb_players.setText("" + players.size());

                    boolean currentPlayerIsInList = false;
                    boolean isEverybodyReady_tmp = true;

                    for (Player player : players) {
                        if (player.getId().equals(currentPlayer.getId())) {
                            currentPlayerIsInList = true;
                        }
                        if (!player.getReady().equals(Utils.SESSION_RDY_YES)){
                            isEverybodyReady_tmp = false;
                        }
                    }

                    if (session_.getState().equals(Utils.SESSION_STATE_LAUNCH_PARTY)){
                        //launchSessionActivity();
                        launchCountDown();
                    }

                    isEverybodyReady = isEverybodyReady_tmp;
                    if (isCreator) {
                        if (isEverybodyReady) {
                            btn_launch_ready.setBackground(ContextCompat.getDrawable(getActivity(), R.drawable.shape_green));
                        } else {
                            btn_launch_ready.setBackground(ContextCompat.getDrawable(getActivity(), R.drawable.shape_orange));
                        }
                    }


                    if (!currentPlayerIsInList) {
                        displayAlertSession("You have been removed from the current session by the creator.");
                        quitSession();
                    }



                } else {
                    displayAlertSession("The current session does not exist anymore.");
                    quitSession();
                }
            }
        });
    }

    private void returnSelectorFragment(){
        SelectorFragment selectorFragment = new SelectorFragment();
        FragmentManager fm = getFragmentManager();
        FragmentTransaction ft = fm.beginTransaction();
        ft.replace(R.id.menu_fl_select, selectorFragment);
        ft.commit();
    }

    private void quitSession(){
        db.updatePlayerNew(session, currentPlayer, Utils.SESSION_LEFT);
        db.removeListenerCurrentSession(session);
        db.removePlayerInSession(session, currentPlayer);
        returnSelectorFragment();
    }

    private void launchSessionActivity() {
        db.removeListenerCurrentSession(session);
        Intent intent = new Intent(getActivity(), SessionActivity.class);
        intent.putExtra("isCreator", isCreator);
        intent.putExtra("session", session);
        intent.putExtra("currentPlayer", currentPlayer);
        startActivity(intent);
        getActivity().finish();
    }

    private void displayAlertSession(String message) {
        AlertDialog alertDialog = new AlertDialog.Builder(getActivity()).create();
        alertDialog.setTitle("Info");
        alertDialog.setMessage(message);
        alertDialog.setButton(AlertDialog.BUTTON_POSITIVE, "OK",
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                });
        alertDialog.show();
    }

    private void makeToast(String text){
        Toast.makeText(getApplicationContext(), text,
                Toast.LENGTH_SHORT).show();
    }

    private void launchCountDown(){
        tv_nb_players_title.setText("Game will begin in ");
        tv_nb_players.setText("5");

        new CountDownTimer(5000,1000) {
            public void onTick(long millisUntilFinished) {
                tv_nb_players.setText(""+(int) millisUntilFinished/1000);
            }

            public void onFinish() {
                tv_nb_players.setText("0");
                launchSessionActivity();
            }

        }.start();
    }
}

File: app/src/main/java/com/example/mlallemant/mentalbattle/ui/menu/fragment/SessionLauncherFragment.kt
package com.example.mlallemant.mentalbattle.ui.menu.fragment

import android.content.Intent
import android.os.Bundle
import android.os.CountDownTimer
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.example.mlallemant.mentalbattle.R
import com.example.mlallemant.mentalbattle.databinding.MenuSessionFragmentBinding
import com.example.mlallemant.mentalbattle.ui.extention.toast
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity.Companion.BUNDLE_EXTRA_CREATOR
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity.Companion.BUNDLE_EXTRA_CURRENT_PLAYER
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity.Companion.BUNDLE_EXTRA_IS_CREATOR
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity.Companion.BUNDLE_EXTRA_NAME
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity.Companion.BUNDLE_EXTRA_PASSWORD
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity.Companion.BUNDLE_EXTRA_SESSION
import com.example.mlallemant.mentalbattle.ui.menu.MenuActivity.OnBackPressedListener
import com.example.mlallemant.mentalbattle.ui.session.SessionActivity
import com.example.mlallemant.mentalbattle.utils.DatabaseManager
import com.example.mlallemant.mentalbattle.utils.Player
import com.example.mlallemant.mentalbattle.utils.Session
import com.example.mlallemant.mentalbattle.utils.Utils
import java.util.*

/**
 * Created by m.lallemant on 15/11/2017.
 */
class SessionLauncherFragment : Fragment() {

    private var _binding: MenuSessionFragmentBinding? = null
    private val binding get() = _binding!!

    //Utils
    private var currentPlayer: Player? = null
    private lateinit var db: DatabaseManager
    private var isCreator = false
    private var name: String? = null
    private var password: String? = null
    private var session: Session? = null
    private var nbPlayer = 0
    private var isReady = false
    private var adapter: PlayersAdapter? = null
    private var isEverybodyReady = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = MenuSessionFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val menuActivity = activity as MenuActivity?
        currentPlayer = menuActivity?.currentPlayer
        isCreator = arguments?.getBoolean(BUNDLE_EXTRA_CREATOR)?:false
        name = arguments?.getString(BUNDLE_EXTRA_NAME)
        password = arguments?.getString(BUNDLE_EXTRA_PASSWORD)
        session = Session(name, password, Utils.SESSION_STATE_WAITING)
        db = DatabaseManager.getInstance()
        db.initListenerCurrentSession(session)
        initUI()
        initListener()
        menuActivity?.setOnBackPressedListener(object : OnBackPressedListener {
            override fun doBack() {
                quitSession()
            }
        })
    }

    override fun onDestroy() {
        super.onDestroy()
        _binding = null
    }

    private fun initUI() {
        with(binding) {
            lvSessionPlayers.visibility = View.VISIBLE
            ivSessionListView.visibility = View.INVISIBLE
            adapter = PlayersAdapter(ArrayList(), session, isCreator, requireContext())
            lvSessionPlayers.adapter = adapter
            adapter?.clear()
            if (isCreator) {
                btnSessionLaunchReady.text = getString(R.string.launch)
            } else {
                btnSessionLaunchReady.text = getString(R.string.ready_question)
            }
            nbPlayer++
            tvSessionNbPlayer.text = nbPlayer.toString()
        }
    }

    private fun initListener() {
        with(binding) {
            ivSessionBack.setOnClickListener {
                onLongClick()
            }
            btnSessionLaunchReady.setOnClickListener {
                onClick()
            }
            db.setOnSessionUpdateListener { session_ ->
                if (session_ != null) {
                    session = session_
                    val players = session_.playerList
                    adapter?.clear()
                    adapter?.addAll(players)
                    tvSessionNbPlayer.text = players.size.toString()
                    var currentPlayerIsInList = false
                    var isEverybodyReadyTmp = true
                    for (player in players) {
                        if (player.id == currentPlayer?.id) {
                            currentPlayerIsInList = true
                        }
                        if (player.ready != Utils.SESSION_RDY_YES) {
                            isEverybodyReadyTmp = false
                        }
                    }
                    if (session_.state == Utils.SESSION_STATE_LAUNCH_PARTY) {
                        launchCountDown()
                    }
                    isEverybodyReady = isEverybodyReadyTmp
                    if (isCreator) {
                        btnSessionLaunchReady.background = when(isEverybodyReady) {
                            true -> ContextCompat.getDrawable(activity!!, R.drawable.shape_green)
                            else -> ContextCompat.getDrawable(activity!!, R.drawable.shape_orange)

                        }
                    }
                    if (!currentPlayerIsInList) {
                        displayAlertSession(getString(R.string.removed_from_session))
                        quitSession()
                    }
                } else {
                    displayAlertSession(getString(R.string.session_does_nor_exist))
                    quitSession()
                }
            }
        }
    }

    private fun onClick() {
        with(binding) {
            if (isCreator) {
                if (isEverybodyReady) {
                    db.updateStateSession(session, Utils.SESSION_STATE_LAUNCH_PARTY)
                } else {
                    toast(getString(R.string.players_not_ready))
                }
            } else {
                btnSessionLaunchReady.apply {
                    if (!isReady) {
                        isReady = true
                        text = getString(R.string.ready_exclamation)
                        background = ContextCompat.getDrawable(activity!!, R.drawable.shape_green)
                        db.updatePlayerReady(session, currentPlayer, Utils.SESSION_RDY_YES)
                    } else {
                        isReady = false
                        text = getString(R.string.ready_question)
                        background = ContextCompat.getDrawable(activity!!, R.drawable.shape_orange)
                        db.updatePlayerReady(session, currentPlayer, Utils.SESSION_RDY_NO)
                    }
                }
            }
        }
    }

    private fun onLongClick() {
        db.removeListenerCurrentSession(session)
        db.insertPlayerInLobby(currentPlayer)
        if (isCreator) {
            db.deleteSession(session)
            returnSelectorFragment()
        } else {
            quitSession()
            db.insertPlayerInLobby(currentPlayer)
        }
    }

    private fun returnSelectorFragment() {
        fragmentManager?.beginTransaction()?.replace(R.id.menu_fl_select, SelectorFragment())?.commit()
    }

    private fun quitSession() {
        db.updatePlayerNew(session, currentPlayer, Utils.SESSION_LEFT)
        db.removeListenerCurrentSession(session)
        db.removePlayerInSession(session, currentPlayer)
        returnSelectorFragment()
    }

    private fun launchSessionActivity() {
        db.removeListenerCurrentSession(session)
        val intent = Intent(activity, SessionActivity::class.java).apply {
            putExtra(BUNDLE_EXTRA_IS_CREATOR, isCreator)
            putExtra(BUNDLE_EXTRA_SESSION, session)
            putExtra(BUNDLE_EXTRA_CURRENT_PLAYER, currentPlayer)
        }
        startActivity(intent)
        activity?.finish()
    }

    private fun displayAlertSession(message: String) {
        AlertDialog.Builder(requireContext()).create().apply {
            setTitle(getString(R.string.alert_session_title))
            setMessage(message)
            setButton(AlertDialog.BUTTON_POSITIVE, getString(R.string.alert_session_btn_ok)) {
                    dialog, _ -> dialog.dismiss()
            }
        }.show()
    }

    private fun launchCountDown() {
        with(binding) {
            tvSessionNbPlayerTitle.text = getString(R.string.launch_count_down_text)
            tvSessionNbPlayer.text = "5"
            object : CountDownTimer(5000, 1000) {
                override fun onTick(millisUntilFinished: Long) {
                    tvSessionNbPlayer.text = (millisUntilFinished.toInt() / 1000).toString()
                }

                override fun onFinish() {
                    tvSessionNbPlayer.text = "0"
                    launchSessionActivity()
                }
            }.start()
        }
    }
}